frameworks/Rust/axum/src/database_mongo.rs

```diff
@@ -1,27 +1,116 @@
 use axum::extract::{Extension, FromRequest, RequestParts};
 use axum::http::StatusCode;
+use futures_util::stream::FuturesUnordered;
+use futures_util::TryStreamExt;
+use std::io;
 
-use async_trait::async_trait;
-use mongodb::{Client, Database};
 use crate::utils::internal_error;
+use crate::{Fortune, World};
+use futures_util::StreamExt;
+use mongodb::bson::doc;
+use mongodb::Database;
 
 pub struct DatabaseConnection(pub Database);
 
 #[async_trait]
 impl<B> FromRequest<B> for DatabaseConnection
-    where
-        B: Send,
+where
+    B: Send,
 {
     type Rejection = (StatusCode, String);
 
     async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
-        let Extension(client) = Extension::<Client>::from_request(req)
+        let Extension(db) = Extension::<Database>::from_request(req)
             .await
             .map_err(internal_error)?;
 
-        let database = client.database("hello_world");
+        Ok(Self(db))
+    }
+}
+
+#[derive(Debug)]
+pub enum MongoError {
+    Io(io::Error),
+    Mongo(mongodb::error::Error),
+}
 
-        Ok(Self(database))
+impl From<io::Error> for MongoError {
+    fn from(err: io::Error) -> Self {
+        MongoError::Io(err)
     }
 }
 
+impl From<mongodb::error::Error> for MongoError {
+    fn from(err: mongodb::error::Error) -> Self {
+        MongoError::Mongo(err)
+    }
+}
+
+pub async fn find_world_by_id(db: Database, id: i32) -> Result<World, MongoError> {
+    let world_collection = db.collection::<World>("world");
+
+    let filter = doc! { "_id": id as f32 };
+
+    let world: World = world_collection
+        .find_one(Some(filter), None)
+        .await
+        .unwrap()
+        .expect("expected world, found none");
+    Ok(world)
+}
+
+pub async fn find_worlds(db: Database, ids: Vec<i32>) -> Result<Vec<World>, MongoError> {
+    let future_worlds = FuturesUnordered::new();
+
+    for id in ids {
+        future_worlds.push(find_world_by_id(db.clone(), id));
+    }
+
+    let worlds: Result<Vec<World>, MongoError> = future_worlds.try_collect().await;
+    worlds
+}
+
+pub async fn fetch_fortunes(db: Database) -> Result<Vec<Fortune>, MongoError> {
+    let fortune_collection = db.collection::<Fortune>("fortune");
+
+    let mut fortune_cursor = fortune_collection
+        .find(None, None)
+        .await
+        .expect("fortunes could not be loaded");
+
+    let mut fortunes: Vec<Fortune> = Vec::with_capacity(100 as usize);
```

Comment:
@dragosv This is the only thing blocking a merge.

> The list data structure must be a dynamic-size or equivalent and should not be dimensioned using foreknowledge of the row-count of the database table.

[Fortune Requirements](https://github.com/TechEmpower/FrameworkBenchmarks/wiki/Project-Information-Framework-Tests-Overview#requirements-3)